File: source/RadXml/RadXmlXmlFile.cpp
#include "RadXmlXmlFile.h"
#include "RadXml.h"

#include "RadXmlApp.h"

#include "RadXmlGame.h"
#include "RadXmlVector3.h"
#include "RadXmlGlobal.h"

#include "languages/language.h"

#include "xml/xml.h"

#include "common/mai_radxml_system.h"

#include "common/com_radxml_strconv.h"

namespace RadXml
{
	XmlFile::XmlFile ()
		: Game (""), App ("")
	{
		init ();
	}

	XmlFile::XmlFile (std::string file)
		: Game (""), App ("")
	{
		init ();
		loadFile (file);
	}

	XmlFile::~XmlFile ()
	{
	}

	void XmlFile::init ()
	{
		dRadXmlVersion = 0;

		filename = "";
		dirPath = "";

		RadXML *ogrGame = RadXML::getSingletonPtr ();

		if (ogrGame->getCurrentXMLFile () == 0)
			ogrGame->setCurrentXMLFile (this);
	}

	bool XmlFile::loadFile (std::string file)
	{
		RadXmlCommon::s13LogMessage (MSG1 + file);

		{
			unsigned int iPos = 0;

			#ifdef defSYS_WIN32
				iPos = file.rfind ("\\");
			#else
				iPos = file.rfind ("/");
			#endif

			filename = file.substr (iPos + 1);
			dirPath = file.substr (0, iPos);
		}

		XMLParser *xmlFile = new XMLParser ();

		if (xmlFile->parseFile (file) == false)
			return (false);

		parseXMLFile (xmlFile);

		delete xmlFile;
		xmlFile = 0;

		return (true);
	}

	void XmlFile::parseXMLFile (XMLParser *xmlFile)
	{
		std::vector<XMLElement *> aryMetas = xmlFile->findAllTags ("meta");

		for (unsigned int iIdx = 0; iIdx < aryMetas.size (); iIdx++)
		{
			XMLElement *xmlElement = aryMetas.at (iIdx);
			std::string strName = xmlElement->getAttributeValue ("name");
			strName = RadXmlCommon::s13ASCII_StringToLower (strName);
			std::string strContent = xmlElement->getAttributeValue ("content");
			strContent = RadXmlCommon::s13ASCII_StringToLower (strContent);

			if (strName == "version")
				dRadXmlVersion = RadXmlCommon::s13StringToDouble (strContent);
		}

		if (dRadXmlVersion == 0.0)
		{
			RadXmlCommon::s13LogMessage (MSG2 + xmlFile->getFilename ());

			return;
		}

		if (dRadXmlVersion < defRADXML_MIN_VERSION)
		{
			RadXmlCommon::s13LogMessage (MSG3 + 
				RadXmlCommon::s13DoubleToString (defRADXML_MIN_VERSION) + MSG4 + xmlFile->getFilename ());

			return;
		}

		if (dRadXmlVersion > defRADXML_VERSION)
		{
			RadXmlCommon::s13LogMessage (MSG5 + 
				RadXmlCommon::s13DoubleToString (defRADXML_VERSION) + MSG4 + xmlFile->getFilename ());

			return;
		}

		std::vector<XMLElement *> aryIncludes = xmlFile->findAllTags ("include");

		for (unsigned int iIdx = 0; iIdx < aryIncludes.size (); iIdx++)
		{
			XMLElement *xmlElement = aryIncludes.at (iIdx);
			std::string strSrc = xmlElement->getAttributeValue ("src");
			std::string strType = xmlElement->getAttributeValue ("type");
			strType = RadXmlCommon::s13ASCII_StringToLower (strType);

			if (strType == "")
				strType = "text/xml";

			if ((strType == "text/xml") || (strType == "xml"))
			{
				strSrc = RadXmlCommon::System_NormalizeDirectoryPath (dirPath + "/" + strSrc, 0, false);
				RadXmlCommon::s13LogMessage (MSG7 + strSrc);

				XMLParser *xmlFile2 = new XMLParser ();

				if (xmlFile2->parseFile (strSrc) == false)
					continue;

				parseXMLFile (xmlFile2);

				delete xmlFile2;
				xmlFile2 = 0;
			}

			if ((strType == "text/javascript") || (strType == "javascript"))
			{
				RadXmlCommon::s13LogMessage (MSG8 + strSrc);

				RadXML::getSingletonPtr ()->compileAndRunJSScript (strSrc);
			}
		}

		std::vector<XMLElement *> aryExternals = xmlFile->findAllTags ("external");

		for (unsigned int iIdx = 0; iIdx < aryExternals.size (); iIdx++)
		{
			XMLElement *xmlElement = aryExternals.at (iIdx);
			std::string strName = xmlElement->getAttributeValue ("name");
			std::string strSrc = xmlElement->getAttributeValue ("src");
			std::string strType = xmlElement->getAttributeValue ("type");
			std::string strOS = xmlElement->getAttributeValue ("os");
			strType = RadXmlCommon::s13ASCII_StringToLower (strType);
			strOS = RadXmlCommon::s13ASCII_StringToLower (strOS);

			#ifdef defSYS_WIN32
				if (strType == "")
					strType = "binary/x86";

				if (strOS == "")
					strOS = "windows";

				if ((strOS == "windows"))
				{
					strSrc = RadXmlCommon::System_NormalizeDirectoryPath (dirPath + "/" + strSrc, 0, false);

					if (loadExternal (strName, strSrc) == true)
						RadXmlCommon::s13LogMessage (MSG9 + strSrc);
					else
						RadXmlCommon::s13LogMessage (MSG10 + strSrc);
				}
			#endif
		}

		std::vector<XMLElement *> aryApp = xmlFile->findAllTags ("app");

		for (unsigned int iIdx = 0; iIdx < aryApp.size (); iIdx++)
		{
			XMLElement *xmlElement = aryApp.at (iIdx);
			std::string strName = xmlElement->getAttributeValue ("name");
			std::string strLoad = xmlElement->getAttributeValue ("load");
			std::string strType = xmlElement->getAttributeValue ("type");
			strType = RadXmlCommon::s13ASCII_StringToLower (strType);
			bool bLoad = true;
			int iType = AT_RENDER;

			if (strLoad != "")
				bLoad = RadXmlCommon::s13StringToBool (strLoad);

			if (strType == "native")
				iType = AT_NATIVE;

			if (strType == "render")
				iType = AT_RENDER;

			if (iIdx == 0)
				parseApp (this, iType, xmlElement);
			else
			{
				App *aApp = new App (strName);

				if (bLoad == true)
					aApp->parseApp (this, iType, xmlElement);
			}
		}

		std::vector<XMLElement *> aryGames = xmlFile->findAllTags ("game");

		for (unsigned int iIdx = 0; iIdx < aryGames.size (); iIdx++)
		{
			XMLElement *xmlElement = aryGames.at (iIdx);
			std::string strName = xmlElement->getAttributeValue ("name");
			std::string strLoad = xmlElement->getAttributeValue ("load");
			bool bLoad = true;

			if (strLoad != "")
				bLoad = RadXmlCommon::s13StringToBool (strLoad);

			if (iIdx == 0)
				parse (this, xmlElement);
			else
			{
				Game *gGame = new Game (strName);

				if (bLoad == true)
					gGame->parse (this, xmlElement);
			}
		}

		std::vector<XMLElement *> aryScripts = xmlFile->findAllTags ("script");
		std::vector<XMLElement *> aryScripts2 = xmlFile->findAllTags ("rxscript");

		for (unsigned int iIdx = 0; iIdx < aryScripts2.size (); iIdx++)
		{
			XMLElement *xmlElement = aryScripts2.at (iIdx);
			aryScripts.push_back (xmlElement);
		}

		for (unsigned int iIdx = 0; iIdx < aryScripts.size (); iIdx++)
		{
			XMLElement *xmlElement = aryScripts.at (iIdx);
			std::string strType = xmlElement->getAttributeValue ("type");
			std::string strSrc = xmlElement->getAttributeValue ("src");
			strType = RadXmlCommon::s13ASCII_StringToLower (strType);

			if (strType == "")
				strType = "text/javascript";

			if (strSrc != "")
			{
				if ((strType == "javascript") || (strType == "text/javascript"))
				{
					strSrc = RadXmlCommon::System_NormalizeDirectoryPath (dirPath + "/" + strSrc, 0, false);
					RadXmlCommon::s13LogMessage (MSG8 + strSrc);
					RadXML::getSingletonPtr ()->compileAndRunJSScript (strSrc);
				}
			}

			if (xmlElement->content != "")
			{
				if ((strType == "javascript") || (strType == "text/javascript"))
					RadXML::getSingletonPtr ()->compileAndRunJSString (xmlElement->content);
			}
		}

		executeJavascriptFromEvent ("onload");
	}

	bool XmlFile::loadExternal (std::string name, std::string filename)
	{
		/// @note Make sure that this binary matches up with the type specified. IE: x86 for x86, ARM for ARM, x64 for x64
		int iLibraryIndex = RadXML::getSingletonPtr ()->loadLibrary (filename);

		if (iLibraryIndex < 0)
			return (false);

		hshDLLs.insert (std::pair<unsigned long, int> (RadXmlCommon::s13StringToHash (name), iLibraryIndex));

		return (true);
	}

	void XmlFile::executeExternalFunction (std::string name, std::string function)
	{
		HASHMAP<unsigned long, int>::iterator itBegin = hshDLLs.find (RadXmlCommon::s13StringToHash (name));
		HASHMAP<unsigned long, int>::iterator itEnd = hshDLLs.end ();

		if (itBegin != itEnd)
		{
			int index = (*itBegin).second;

			#ifndef defAPI_WXWIDGETS
			#ifdef defSYS_WIN32
				RadXML::getSingletonPtr ()->getDLLFunction (index, function) ();
			#endif
			#else
				((RadXmlFunction)RadXML::getSingletonPtr ()->getDLLFunction (index, function)) ();
			#endif
		}
	}

	bool XmlFile::unloadExternal (std::string name)
	{
		HASHMAP<unsigned long, int>::iterator itBegin = hshDLLs.find (RadXmlCommon::s13StringToHash (name));
		HASHMAP<unsigned long, int>::iterator itEnd = hshDLLs.end ();

		if (itBegin != itEnd)
		{
			int index = (*itBegin).second;
			bool bReturn = RadXML::getSingletonPtr ()->unloadLibrary (index);

			return (bReturn);
		}

		return (false);
	}

	void XmlFile::processDeleteObjects ()
	{
		processDeleteAppObjects ();
	}
}

File: html5/RadXml/RadXmlXmlFile.js
/*
	MIT-LICENSE
	Copyright (c) 2013 Higher Edge Software, LLC

	Permission is hereby granted, free of charge, to any person obtaining a copy of this software 
	and associated documentation files (the "Software"), to deal in the Software without restriction, 
	including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, 
	and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
	subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all copies or substantial 
	portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT 
	LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
	IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION 
	WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

var ApplicationType = 
{
	AT_DESKTOP: 1, 
	AT_MOBILE: 2
};

function XmlFile ()
{
	this._dRadXmlVersion = 0;
	this._filename = "";
	this._dirPath = "";
	this._type = ApplicationType.AT_DESKTOP;
	this._tempXmlFile = null;
	this._numIncludesToLoad = 0;
	this._doingAsyncScriptLoading = false;

	this.loadFile = function (file)
	{
		var iPos = file.lastIndexOf ("/");

		this._filename = file.substr (iPos + 1);
		this._dirPath = file.substr (0, iPos);

		var strTemp = this._dirPath + "/" + this._filename;
		file = strTemp.normalizeDirectoryPath (0, false);

		logMessage (RadXmlLanguage.MSG20 + file);

		var strXML = RadXML._getSyncResponseFromHTTP (file);
		var xmlFile = this._openXMLFile (strXML, file);
		this.parseXMLFile (xmlFile);
	}

	this.loadString = function (xmlString)
	{
		var file = window.location.href;
		var iPos = file.lastIndexOf ("/");

		this._filename = file.substr (iPos + 1);
		this._dirPath = file.substr (0, iPos);

		var strTemp = this._dirPath + "/" + this._filename;
		file = strTemp.normalizeDirectoryPath (0, false);

		logMessage (RadXmlLanguage.MSG39);

		var xmlFile = this._openXMLFile (xmlString, file);
		this.parseXMLFile (xmlFile);
	}

	this._openXMLFile = function (xmlString, file)
	{
		var xmlFile = null;

		if (DOMParser)
		{
			var xmlParser = new DOMParser ();
			xmlFile = xmlParser.parseFromString (xmlString, "text/xml");
		}
		else
		{
			xmlFile = new ActiveXObject ("Microsoft.XMLDOM");
			xmlFile.async = false;
			xmlFile.loadXML (xmlString);
		}

		xmlFile.fileName = file;

		return (xmlFile);
	}

	this.parseXMLFile = function (xmlFile)
	{
		var iTempIncludes = 0;
		var aryMetas = xmlFile.getElementsByTagName ("meta");
		this._tempXmlFile = xmlFile;

		for (var iIdx = 0; iIdx < aryMetas.length; iIdx++)
		{
			var xmlElement = aryMetas[iIdx];
			var strName = RadXML._getAttribute (xmlElement, "name").value;
			strName = strName.toLowerCase ();
			var strContent = RadXML._getAttribute (xmlElement, "content").value;
			strContent = strContent.toLowerCase ();

			if (strName == "version")
				this._dRadXmlVersion = RadXML._parseFloat (strContent);

			if (strName == "collisiondetection")
				Game.setCollisionDetection (parseBoolean (strContent));

			if (strName == "type")
			{
				var iType = ApplicationType.AT_DESKTOP;

				if (strContent == "desktop")
					iType = ApplicationType.AT_DESKTOP;

				if (strContent == "mobile")
					iType = ApplicationType.AT_MOBILE;

				this._type = iType;
			}
		}

		if (this._dRadXmlVersion == 0.0)
		{
			logMessage (RadXmlLanguage.MSG21 + xmlFile.fileName);

			return;
		}

		if (this._dRadXmlVersion < RadXML.defRADXML_MIN_VERSION)
		{
			logMessage (RadXmlLanguage.MSG22 + defRADXML_MIN_VERSION + 
						RadXmlLanguage.MSG23 + xmlFile.fileName);

			return;
		}

		if (this._dRadXmlVersion > RadXML.defRADXML_VERSION)
		{
			logMessage (RadXmlLanguage.MSG24 + RadXML.defRADXML_VERSION + 
					RadXmlLanguage.MSG23 + xmlFile.fileName);

			return;
		}

		var aryIncludes = xmlFile.getElementsByTagName ("include");

		for (var iIdx = 0; iIdx < aryIncludes.length; iIdx++)
		{
			var xmlElement = aryIncludes[iIdx];
			var strSrc = RadXML._getAttribute (xmlElement, "src").value;
			var strType = RadXML._getAttribute (xmlElement, "type").value;
			strType = strType.toLowerCase ();

			if (strType == "")
				strType = "text/xml";

			if ((strType == "text/xml") || (strType == "xml"))
			{
				var strTemp = this._dirPath + "/" + strSrc;
				strSrc = strTemp.normalizeDirectoryPath (0, false);
				s13LogMessage (RadXmlLanguage.MSG25 + strSrc);

				var strXML = RadXML._getSyncResponseFromHTTP (strSrc);
				var xmlFile = this._openXMLFile (strXML, strSrc);
				this.parseXMLFile (xmlFile);

				delete xmlFile;
			}

			if ((strType == "text/javascript") || (strType == "javascript"))
			{
				logMessage (RadXmlLanguage.MSG26 + strSrc);

				RadXML.compileAndRunJSScript (strSrc);
				iTempIncludes++;
			}
		}

		var aryScripts = xmlFile.getElementsByTagName ("script");
		var aryScripts2 = xmlFile.getElementsByTagName ("rxscript");

		for (var iSdx = 0; iSdx < 2; iSdx++)
		{
			var aryS = null;

			if (iSdx == 0)
				aryS = aryScripts;

			if (iSdx == 1)
				aryS = aryScripts2;

			for (var iIdx = 0; iIdx < aryS.length; iIdx++)
			{
				var xmlElement = aryS[iIdx];
				var strType = RadXML._getAttribute (xmlElement, "type").value;
				var strSrc = RadXML._getAttribute (xmlElement, "src").value;
				strType = strType.toLowerCase ();

				if (strType == "")
					strType = "text/javascript";

				if (strSrc != "")
				{
					if ((strType == "javascript") || (strType == "text/javascript"))
					{
						var strTemp = this._dirPath + "/" + strSrc;
						strSrc = strTemp.normalizeDirectoryPath (0, false);
						logMessage (RadXmlLanguage.MSG26 + strSrc);
						RadXML.compileAndRunJSScript (strSrc);
						iTempIncludes++;
					}
				}

				if (xmlElement.textContent != "")
				{
					if ((strType == "javascript") || (strType == "text/javascript"))
					{
						include ("", "", xmlElement.textContent);
						iTempIncludes++;
					}
				}
			}
		}

		var aryExternals = xmlFile.getElementsByTagName ("external");

		for (var iIdx = 0; iIdx < aryExternals.length; iIdx++)
		{
			var xmlElement = aryExternals[iIdx];
			var strName = RadXML._getAttribute (xmlElement, "name").value;
			var strSrc = RadXML._getAttribute (xmlElement, "src").value;
			var strType = RadXML._getAttribute (xmlElement, "type").value;
			var strOS = RadXML._getAttribute (xmlElement, "os").value;
			strType = strType.toLowerCase ();
			strOS = strOS.toLowerCase ();

			if (strType == "")
				strType = "binary/x86";

			if (strOS == "")
				strOS = "windows";

			logMessage (MSG27);
		}

		this._numIncludesToLoad = iTempIncludes;

		if ((this._numIncludesToLoad == 0) || (this._doingAsyncScriptLoading == false))
		{
			this.continueLoadingXml ();
			this._tempXmlFile = null;
		}
	}

	this.checkIfScriptsAreLoaded = function ()
	{
		if (this._numIncludesToLoad <= 0)
		{
			this.continueLoadingXml ();
			this._tempXmlFile = null;
		}
	}

	this.continueLoadingXml = function ()
	{
		var xmlFile = this._tempXmlFile;
		var aryApp = xmlFile.getElementsByTagName ("app");

		for (var iIdx = 0; iIdx < aryApp.length; iIdx++)
		{
			var xmlElement = aryApp[iIdx];
			var strName = RadXML._getAttribute (xmlElement, "name").value;
			var strLoad = RadXML._getAttribute (xmlElement, "load").value;
			var strType = RadXML._getAttribute (xmlElement, "type").value;
			var strOnLoad = RadXML._getAttribute (xmlElement, "onload").value;
			strType = strType.toLowerCase ();
			var bLoad = true;
			var iType = AppTypes.AT_RENDER;

			if (strLoad != "")
				bLoad = parseBoolean (strLoad);

			if (strType == "native")
				iType = AppTypes.AT_NATIVE;

			if (strType == "render")
				iType = AppTypes.AT_RENDER;

			if (bLoad == true)
			{
				if (iIdx == 0)
					this.app._parseApp (this, iType, xmlElement);
				else
				{
					var aApp = new App (strName);
					aApp._parseApp (this, iType, xmlElement);
				}

				RadXML.compileAndRunJSString (strOnLoad);
			}
		}

		var aryGames = xmlFile.getElementsByTagName ("game");

		for (var iIdx = 0; iIdx < aryGames.length; iIdx++)
		{
			var xmlElement = aryGames[iIdx];
			var strName = RadXML._getAttribute (xmlElement, "name").value;
			var strLoad = RadXML._getAttribute (xmlElement, "load").value;
			var bLoad = true;

			if (strLoad != "")
				bLoad = parseBoolean (strLoad);

			if (bLoad == true)
			{
				if (iIdx == 0)
					this.game._parse (this, xmlElement);
				else
				{
					var gGame = new Game (strName);
					gGame._parse (this, xmlElement);
				}

				RadXML.compileAndRunJSString (strOnLoad);
			}
		}
	}

	this.getMovable = function (name)
	{
		return (this.game.getMovableByName (name));
	}
}

XmlFile.prototype.app = App;
XmlFile.prototype.game = Game;

